src/AspNet.Security.OAuth.Introspection/OAuthIntrospectionHandler.cs

```diff
@@ -23,59 +23,138 @@
+                    }
+                    if (createTicketContext.Skipped) {
+                        return AuthenticateResult.Skip();
+                    }
+
+                    // Create a new authentication ticket from the introspection
+                    // response returned by the authorization server.
+                    ticket = createTicketContext.Ticket ?? await CreateTicketAsync(payload);
+                    Debug.Assert(ticket != null);
+
+                    await StoreTicketAsync(token, ticket);
+                }
+
+                // Allow for interception and handling of the token validated event.
+                var tokenValidatedContext = new TokenValidatedContext(Context, Options, ticket);
+                await Options.Events.TokenValidated(tokenValidatedContext);
+                if (tokenValidatedContext.HandledResponse)
+                {
+                    return AuthenticateResult.Success(tokenValidatedContext.Ticket);
                 }
+                if (tokenValidatedContext.Skipped)
+                {
+                    Logger.LogInformation("Authentication was skipped by event processing.");
 
-                // Ensure that the access token was issued
-                // to be used with this resource server.
-                if (!await ValidateAudienceAsync(payload)) {
-                    return AuthenticateResult.Fail("Authentication failed because the access token " +
-                                                   "was not valid for this resource server.");
+                    return AuthenticateResult.Skip();
                 }
 
-                // Create a new authentication ticket from the introspection
-                // response returned by the authorization server.
-                ticket = await CreateTicketAsync(payload);
-                Debug.Assert(ticket != null);
+                // Flow the ticket changes
+                ticket = tokenValidatedContext.Ticket;
 
-                await StoreTicketAsync(token, ticket);
-            }
+                // Ensure that the authentication ticket is still valid.
+                if (ticket.Properties.ExpiresUtc.HasValue &&
+                    ticket.Properties.ExpiresUtc.Value < Options.SystemClock.UtcNow) {
+                    return AuthenticateResult.Fail("Authentication failed because the access token was expired.");
+                }
 
-            // Ensure that the authentication ticket is still valid.
-            if (ticket.Properties.ExpiresUtc.HasValue &&
-                ticket.Properties.ExpiresUtc.Value < Options.SystemClock.UtcNow) {
-                return AuthenticateResult.Fail("Authentication failed because the access token was expired.");
+                return AuthenticateResult.Success(ticket);
             }
+            catch (Exception exception)
+            {
+                var authenticationFailedContext = new AuthenticationFailedContext(Context, Options) {
```

Comment:
For a few reasons:
  - Catching all the exception types in such a big block is an anti-pattern :smile: 
  - This event proved to be nasty in the JWT bearer middleware because people tried to use it to write something to the response stream, which is baaaaaaaaad since we're outside the normal request process at this stage (we can't stop the request processing from `HandleAuthenticateAsync`).
  - The validation middleware shouldn't throw at all (just like Katana's OAuth2 bearer middleware actually). Any exception thrown from `HandleAuthenticateAsync` would indicate an error in our code or an error that was not expected at all, not really something caused by an incorrect authentication. This is also true for the introspection middleware, though the use of `HttpClient` is more likely to throw transient exceptions.